### ISSUE TITLE
pip installation failed on Ubuntu

### ISSUE THREAD

**Issue by Co1lin** [BQSKit/bqskit#111]
Hi! I tried to install this package on a Ubuntu machine, but it failed.

System information:

```python
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.1 LTS
Release:        22.04
Codename:       jammy

Linux version 5.15.0-1026-aws (buildd@lcy02-amd64-105) (gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0, GNU ld (GNU Binutils for Ubuntu) 2.38) #30-Ubuntu SMP Wed Nov 23 14:15:21 UTC 2022

Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         48 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  4
  On-line CPU(s) list:   0-3
Vendor ID:               AuthenticAMD
  Model name:            AMD EPYC 7R32
...
```

Logs:

```python
$ pip install 'bqskit[ext]'
...
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. See https://pip.pypa.io/warnings/backtracking for guidance. If you want to abort this run, press Ctrl + C.
  Downloading matplotlib-3.3.3.tar.gz (37.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.9/37.9 MB 67.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Downloading matplotlib-3.3.2.tar.gz (37.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.9/37.9 MB 23.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Downloading matplotlib-3.3.1.tar.gz (38.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.8/38.8 MB 64.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Downloading matplotlib-3.3.0.tar.gz (38.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.8/38.8 MB 63.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Downloading matplotlib-3.2.2.tar.gz (40.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.3/40.3 MB 65.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Downloading matplotlib-3.2.1.tar.gz (40.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.3/40.3 MB 63.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Downloading matplotlib-3.2.0.tar.gz (40.5 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.5/40.5 MB 63.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Downloading matplotlib-3.1.3.tar.gz (40.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.9/40.9 MB 57.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Downloading matplotlib-3.1.2.tar.gz (40.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.9/40.9 MB 59.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Downloading matplotlib-3.1.1.tar.gz (37.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.8/37.8 MB 64.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Downloading matplotlib-3.1.0.tar.gz (37.2 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.2/37.2 MB 64.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Downloading matplotlib-3.0.3.tar.gz (36.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 36.6/36.6 MB 67.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [52 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-m6b3vzlg/matplotlib_616bfc99eb4e49e09a0cb98e61fae23c/setup.py", line 225, in <module>
          msg = pkg.install_help_msg()
        File "/tmp/pip-install-m6b3vzlg/matplotlib_616bfc99eb4e49e09a0cb98e61fae23c/setupext.py", line 650, in install_help_msg
          release = platform.linux_distribution()[0].lower()
      AttributeError: module 'platform' has no attribute 'linux_distribution'
      IMPORTANT WARNING:
          pkg-config is not installed.
          matplotlib may not be able to find some of its dependencies
      ============================================================================
      Edit setup.cfg to change the build options
      
      BUILDING MATPLOTLIB
                  matplotlib: yes [3.0.3]
                      python: yes [3.10.8 (main, Nov 24 2022, 14:13:03) [GCC
                              11.2.0]]
                    platform: yes [linux]
      
      REQUIRED DEPENDENCIES AND EXTENSIONS
                       numpy: yes [not found. pip may install it below.]
            install_requires: yes [handled by setuptools]
                      libagg: yes [pkg-config information for 'libagg' could not
                              be found. Using local copy.]
                    freetype: no  [The C/C++ header for freetype2 (ft2build.h)
                              could not be found.  You may need to install the
                              development package.]
                         png: no  [pkg-config information for 'libpng' could not
                              be found.]
                       qhull: yes [pkg-config information for 'libqhull' could not
                              be found. Using local copy.]
      
      OPTIONAL SUBPACKAGES
                 sample_data: yes [installing]
                    toolkits: yes [installing]
                       tests: no  [skipping due to configuration]
              toolkits_tests: no  [skipping due to configuration]
      
      OPTIONAL BACKEND EXTENSIONS
                         agg: yes [installing]
                       tkagg: yes [installing; run-time loading from Python Tcl /
                              Tk]
                      macosx: no  [Mac OS-X only]
                   windowing: no  [Microsoft Windows only]
      
      OPTIONAL PACKAGE DATA
                        dlls: no  [skipping due to configuration]
      
      ============================================================================
                              * The following required packages can not be built:
                              * freetype, png
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

**Comment by edyounis** [comment]
I am not sure if they are related, but I am also experiencing an issue with the installation taking a long time. While I am looking into this, can you try the following command and let me know if it resolves your issue:

```bash
sudo apt-get install libfreetype6-dev pkg-config libpng16-devtools
```

**Comment by Co1lin** [comment]
Hi @edyounis! I tried to run `pip install 'bqskit[ext]'` several times, and it always couldn't finish. `pip` gave some warning that it tried to find a compatible solution for all version constraints required by dependent packages, but it seemed hard to do so. I think `pip install` took a long time because it got stuck in some search algorithms, such as DFS, for different version combinations.

It seems that `pip install bqskit` can be quickly finished, so could you try to verify this? If it's right, then you can try to install other dependent quantum toolkits one by one to see which one has some problem, I guess.

**Comment by Co1lin** [comment]
I tried your command above, but made a modification because `libpng16-devtools` is not available on my system, so I changed it to `libpng-dev` according to the tips given by `apt`. However, I still cannot finish `pip install 'bqskit[ext]'` as it kept trying multiple versions of the packages.

**Comment by edyounis** [comment]
Thanks for checking the `apt` command. Yeah, it's a bad interaction between cirq and dask. If you need an immediate fix, you can remove all occurrences of cirq from BQSKit, and it should work fine (unless you use cirq :)).

The proper way to handle this long-term is to redo how the `ext` subpackage is imported. To import anything from `bqskit.ext` you must import everything. This is obviously bad and was only meant to be temporary. The right thing to do is not have any of the frameworks installed by pip and when you import something from `bqskit.ext` check for the specific package. If you find the package, great! Otherwise, we should print a pretty message explaining to the user to install the proper package and how to do it.

It would be greatly appreciated if you could make the change and a PR for this, otherwise, I will fix this in the upcoming 1.0.4 update.

**Comment by Co1lin** [comment]
As a workaround, I just ran `pip install --no-deps` and manually installed every missing package. I will try the new version later. Thanks for the instant fix!

**Comment by edyounis** [comment]
This should be fixed in the 1.0.4 release, feel free to re-open if it comes up again
